code/lib/preview-api/src/modules/store/StoryStore.ts

```diff
@@ -143,22 +143,38 @@ export class StoryStore<TRenderer extends Renderer> {
     );
   }
 
-  loadAllCSFFiles(): Promise<StoryStore<TRenderer>['cachedCSFFiles']> {
+  loadAllCSFFiles({ batchSize = EXTRACT_BATCH_SIZE } = {}): Promise<
+    StoryStore<TRenderer>['cachedCSFFiles']
+  > {
     if (!this.storyIndex) throw new Error(`loadAllCSFFiles called before initialization`);
 
-    const importPaths: Record<Path, StoryId> = {};
-    Object.entries(this.storyIndex.entries).forEach(([storyId, { importPath }]) => {
-      importPaths[importPath] = storyId;
-    });
+    const importPaths = Object.entries(this.storyIndex.entries).map(([storyId, { importPath }]) => [
+      importPath,
+      storyId,
+    ]);
+
+    const loadInBatches = (
+      remainingImportPaths: typeof importPaths
+    ): Promise<{ importPath: Path; csfFile: CSFFile<TRenderer> }[]> => {
+      if (remainingImportPaths.length === 0) return SynchronousPromise.resolve([]);
+
+      const csfFilePromiseList = remainingImportPaths
+        .slice(0, batchSize)
+        .map(([importPath, storyId]) =>
+          this.loadCSFFileByStoryId(storyId).then((csfFile) => ({
+            importPath,
+            csfFile,
+          }))
+        );
 
-    const csfFilePromiseList = Object.entries(importPaths).map(([importPath, storyId]) =>
-      this.loadCSFFileByStoryId(storyId).then((csfFile) => ({
-        importPath,
-        csfFile,
-      }))
-    );
+      return SynchronousPromise.all(csfFilePromiseList).then((firstResults) =>
+        loadInBatches(remainingImportPaths.slice(batchSize)).then((restResults) =>
+          firstResults.concat(restResults)
+        )
+      );
+    };
```

Comment:
There's a package called `p-limit` that I've used before that's slightly more sophisticated:
https://github.com/storybookjs/storybook/blob/5a557d37ce573e15290cd7ed43ba434e3b42e6b5/scripts/run-registry.ts#L11
It doesn't just do batches, but manages a pool of workers.

Not saying we should change, but if we care about this running as fast as it possibly could, then this isn't the fastest implementation.

The implementation will do a batch of n.
If 1 out of those n takes 100 second, the whole batch will take 100s.

I'm not sure that matters in this code-path.

Reply:
Thanks @ndelangen. 

We can't really use such tools in this code path because of the requirement to use `SynchronousPromise`s as our underlying mechanism (which is for Storyshot's sake).

I think you are right that it probably doesn't really matter for this code path, which should be pretty fast either way, and there probably isn't going to be a heap of variance between different CSF files.

Still it would be better to use something like `p-limit`. Perhaps once we drop SS support we could change to it.